task_processor/decorators.py

```diff
@@ -0,0 +1,195 @@
+import logging
+import os
+import typing
+from datetime import datetime, time, timedelta
+from inspect import getmodule
+from threading import Thread
+
+from django.conf import settings
+from django.db.transaction import on_commit
+from django.utils import timezone
+
+from task_processor.exceptions import InvalidArgumentsError, TaskQueueFullError
+from task_processor.models import RecurringTask, Task, TaskPriority
+from task_processor.task_registry import register_task
+from task_processor.task_run_method import TaskRunMethod
+
+P = typing.ParamSpec("P")
+
+logger = logging.getLogger(__name__)
+
+
+class TaskHandler(typing.Generic[P]):
+    __slots__ = (
+        "unwrapped",
+        "queue_size",
+        "priority",
+        "transaction_on_commit",
+        "task_identifier",
+    )
+
+    unwrapped: typing.Callable[P, None]
+
+    def __init__(
+        self,
+        f: typing.Callable[P, None],
+        *,
+        task_name: str | None = None,
+        queue_size: int | None = None,
+        priority: TaskPriority = TaskPriority.NORMAL,
+        transaction_on_commit: bool = True,
+    ) -> None:
+        self.unwrapped = f
+        self.queue_size = queue_size
+        self.priority = priority
+        self.transaction_on_commit = transaction_on_commit
+
+        task_name = task_name or f.__name__
+        task_module = getmodule(f).__name__.rsplit(".")[-1]
+        self.task_identifier = task_identifier = f"{task_module}.{task_name}"
+        register_task(task_identifier, f)
+
+    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> None:
+        _validate_inputs(*args, **kwargs)
+        return self.unwrapped(*args, **kwargs)
+
+    def delay(
+        self,
+        *,
+        delay_until: datetime | None = None,
+        # TODO @khvn26 consider typing `args` and `kwargs` with `ParamSpec`
+        # (will require a change to the signature)
+        args: tuple[typing.Any, ...] = (),
+        kwargs: dict[str, typing.Any] | None = None,
+    ) -> Task | None:
+        logger.debug("Request to run task '%s' asynchronously.", self.task_identifier)
+
+        kwargs = kwargs or {}
+
+        if delay_until and settings.TASK_RUN_METHOD != TaskRunMethod.TASK_PROCESSOR:
+            logger.warning(
+                "Cannot schedule tasks to run in the future without task processor."
+            )
+            return
```

Comment:
Should this really silently fail?

Reply:
That's a good question. I think the original thought process (since this code hasn't changed) was that this could happen at runtime in an environment which doesn't have the task processor running. In practice, I believe that the only tasks that we schedule for the future are those related to the edge api only. I've added a TODO for now.